Week3 - Mohammed Saif ( 4 exercises )

MoSaif00 opened the pull request:
Within the week3, I have worked on 4 exercises :

1. Normalization -> In this exercise, I have a table and 3 questions that I answered according to the table data. 
2. Transaction  -> In the transaction exercise, I built 3 files (  transaction create tables, transaction insert values and transactions,), in the first file I created the required tables in the guide, the second file I inserted some values, then with the third files I wrote some queries but with using transactions method. 
3. SQL Injection  -> this exercise was SQL Injection where the code could take advantage of SQL injection while retrieving the data then I rewrote the code so it can prevent the SQL injections attacks.
4. MongoDB CRUD -> In this exercise, I practised the concept of MongoDB CRUD.

mogaleaf commented on Week3/homework/1-Normalization/1-Normalization.js:
```diff
@@ -0,0 +1,51 @@
+/*
+The columns that violate 1NF are : 
+- member_id => there is a duplicate records and primary key is not applied. 
+- member_address => the data is not atomic because we should have only single value in a column. 
```
I m not sure I see a problem in this column

MoSaif00 replied:
You are totally right, as I notice, it is just street name and number, there is no city name or zip codes that could make it violating 1NF. Thanks for your time

mogaleaf commented on Week3/homework/1-Normalization/1-Normalization.js:
```diff
@@ -0,0 +1,51 @@
+// What columns violate 1NF?
+/*
+The columns that violate 1NF are : 
+- member_id => there is a duplicate records and primary key is not applied. 
```
True but its not 1NF

mogaleaf commented on Week3/homework/1-Normalization/1-Normalization.js:
```diff
@@ -0,0 +1,51 @@
+The columns that violate 1NF are : 
+- member_id => there is a duplicate records and primary key is not applied. 
+- member_address => the data is not atomic because we should have only single value in a column. 
+- dinner_data => the data are date type but not the same kind and domains, sometimes will not be easy to retrieve data by dinner_date because the format of each record is different and not unique.
```
yes terrible

mogaleaf commented on Week3/homework/1-Normalization/1-Normalization.js:
```diff
@@ -0,0 +1,51 @@
+- member_id => there is a duplicate records and primary key is not applied. 
+- member_address => the data is not atomic because we should have only single value in a column. 
+- dinner_data => the data are date type but not the same kind and domains, sometimes will not be easy to retrieve data by dinner_date because the format of each record is different and not unique.
+- food_code => the data is not atomic and the column has multiple values which even we split them to different tables it will violating 1NF.
```
yes

mogaleaf commented on Week3/homework/1-Normalization/1-Normalization.js:
```diff
@@ -0,0 +1,51 @@
+- member_address => the data is not atomic because we should have only single value in a column. 
+- dinner_data => the data are date type but not the same kind and domains, sometimes will not be easy to retrieve data by dinner_date because the format of each record is different and not unique.
+- food_code => the data is not atomic and the column has multiple values which even we split them to different tables it will violating 1NF.
+- food_description => the data is not atomic and the column has multiple values.
```
yes

mogaleaf commented on Week3/homework/1-Normalization/1-Normalization.js:
```diff
@@ -0,0 +1,51 @@
+- Table (members) => member_id(primary key), member_name, member_house_number, member_street_name.
+- Table (venues) => venue_code(primary key), venue_description. 
+- Table (foods) => food_code(primary key), food_description.
+- Table (dinners) => 
```
Are you sure you can fit all of this in one table ? What if one dinner has multiple members ?

mogaleaf commented on Week3/homework/1-Normalization/1-Normalization.js:
```diff
@@ -0,0 +1,51 @@
+    member_id(foreign key References members.member_id),
+    venue_code(foreign key References venues.venue_code),
+    food_code(foreign key References foods.food_code).
+    (in dinner table we could add many columns and called food1 , food2, food3 ..etc , because we might find it still violating 1NF if we have multiple values in same column. is there any suggestion to solve this ?! )
```
yes, you can have like last week a relation tables :
 dinner_id, member_id, food_code that link the dinner,member and what he chose to eat.

MoSaif00 replied:
Thank you @mogaleaf for your time reviewing my homework, at beginning I thought that making a dinner table could make a relationship between other tables, but after your suggestion, made another relation table called dinner_member that hold member_id, dinner_id and food code. but I have a question do you think is right to references venue_code within the foods table, because as I noticed that food and venue share the same values!

mogaleaf commented on Week3/homework/2-Transactions/transaction.js:
```diff
@@ -0,0 +1,47 @@
+const util = require('util');
```
Good

mogaleaf commented on Week3/homework/3-SQL-ingection/3-SQL_injection.js:
```diff
@@ -0,0 +1,53 @@
+function getPopulation(country, name, code) {
+  // assuming that connection to the database is established and stored as conn
+  connection.query(
+    `SELECT Population FROM ${country} WHERE Name = ? AND code = ?`,
```
You need to escape the table name too otherwise you can also inject here.

mogaleaf commented on Week3/homework/4-MongoDB_CRUD/4-MongoDB_create.js:
```diff
@@ -0,0 +1,36 @@
+const {MongoClient} = require('mongodb');
```
Good for the mongodb

mogaleaf reviewed (commented):
Hi,

Good homework, you need to review a but exercise 1, the last part. Try to really see how you would do to organize a dinner with multiple members eating different type of food.
The transaction exercise and mongo db are good.
You need to escape also table name in order to be safe in the injection exercise.
thanks !

MoSaif00 commented:
> Hi,
> 
> Good homework, you need to review a but exercise 1, the last part. Try to really see how you would do to organize a dinner with multiple members eating different type of food.
> The transaction exercise and mongo db are good.
> You need to escape also table name in order to be safe in the injection exercise.
> thanks !

Thank you @mogaleaf  for your time and effort reviewing my homework, I rewrote and edited the tables in the first exercise according to your suggestion. updated the Sql-injection exercise as well.

Thank you

mogaleaf commented:
Super thanks for the modifications !
